src/utils.ts

```diff
@@ -0,0 +1,96 @@
+type CustomPayload = {
+  website: string;
+  [key: string]: unknown;
+};
+
+type TrackCallback = (props: {
+  hostname: string;
+  language: string;
+  referrer: string;
+  screen: string;
+  title: string;
+  url: string;
+  website: string;
+}) => { website: string; [key: string]: unknown };
+
+type TrackingOptions = {
+  pathname?: string;
+  processUrl?: (url: string) => string;
+};
+
+export const getTrackFunction = function <
+  TEventNames extends readonly string[],
+  TEventData extends Partial<
+    Record<TEventNames[number], { [key: string]: unknown }>
+  >,
+>(eventNames: TEventNames, trackingOptions?: TrackingOptions) {
+  // These definitions are based on https://www.npmjs.com/package/@types/umami
+  // however, note that these are created with generics so we get intellisense on the events defined
```

Comment:
Should we send a PR to @types/umami?

Reply:
I don't think they will accept that, as they are typed as a generic "string", accepting anything.  The benefit of this override is that you get type-safety and intellisense (but it requires a bit of work on the js side too - so sending the type changes to Umami won't be enough)